ql/src/java/org/apache/hadoop/hive/ql/exec/vector/mapjoin/VectorMapJoinAntiJoinLongOperator.java

```diff
@@ -0,0 +1,315 @@
+        }
+        finishAntiRepeated(batch, joinResult, hashSetResults[0]);
+      } else {
+        // NOT Repeating.
+
+        if (LOG.isDebugEnabled()) {
+          LOG.debug(CLASS_NAME + " batch #" + batchCounter + " non-repeated");
+        }
+
+        // We remember any matching rows in matches / matchSize.  At the end of the loop,
+        // selected / batch.size will represent both matching and non-matching rows for outer join.
+        // Only deferred rows will have been removed from selected.
+        int selected[] = batch.selected;
+        boolean selectedInUse = batch.selectedInUse;
+
+        int hashSetResultCount = 0;
+        int allMatchCount = 0;
+        int spillCount = 0;
+        long saveKey = 0;
+
+        // We optimize performance by only looking up the first key in a series of equal keys.
+        boolean haveSaveKey = false;
+        JoinUtil.JoinResult saveJoinResult = JoinUtil.JoinResult.NOMATCH;
+
+        // Logical loop over the rows in the batch since the batch may have selected in use.
+        for (int logical = 0; logical < inputLogicalSize; logical++) {
+          int batchIndex = (selectedInUse ? selected[logical] : logical);
+
+          // Single-Column Long get key.
+          long currentKey;
+          boolean isNull;
+          if (!joinColVector.noNulls && joinColVector.isNull[batchIndex]) {
+            currentKey = 0;
+            isNull = true;
+          } else {
+            currentKey = vector[batchIndex];
+            isNull = false;
+          }
+
+          // Equal key series checking.
+          if (isNull || !haveSaveKey || currentKey != saveKey) {
+            // New key.
+            if (haveSaveKey) {
+              // Move on with our counts.
+              switch (saveJoinResult) {
+              case MATCH:
+                // We have extracted the existence from the hash set result, so we don't keep it.
+                break;
+              case SPILL:
+                // We keep the hash set result for its spill information.
+                hashSetResultCount++;
+                break;
+              case NOMATCH:
+                break;
+              }
+            }
+
+            if (isNull) {
+              saveJoinResult = JoinUtil.JoinResult.MATCH;
+              haveSaveKey = false;
+            } else {
+              // Regardless of our matching result, we keep that information to make multiple use
+              // of it for a possible series of equal keys.
+              haveSaveKey = true;
+              saveKey = currentKey;
+              if (useMinMax && (currentKey < min || currentKey > max)) {
+                // Key out of range for whole hash table, is a valid match for anti join.
+                saveJoinResult = JoinUtil.JoinResult.NOMATCH;
+              } else {
+                saveJoinResult = hashSet.contains(currentKey, hashSetResults[hashSetResultCount]);
+              }
+
+              // Reverse the match result for anti join.
```

Comment:
Again this could be replaced with an Inversion utility function

Reply:
done